src/os/src/windows/os_platform_ifaddrs.c

```diff
@@ -0,0 +1,260 @@
+        assert(ret != ERROR_INVALID_PARAMETER);
+        switch (ret) {
+            case ERROR_BUFFER_OVERFLOW:
+                err = ERANGE;
+                os_free(buf);
+                if ((buf = (IP_ADAPTER_ADDRESSES *)os_malloc(bufsz)) == NULL) {
+                    err = ENOMEM;
+                }
+                break;
+            case ERROR_NOT_ENOUGH_MEMORY:
+                err = ENOMEM;
+                break;
+            case ERROR_SUCCESS:
+            case ERROR_ADDRESS_NOT_ASSOCIATED: /* No address associated yet. */
+            case ERROR_NO_DATA: /* No adapters that match the filter. */
+            default:
+                err = 0;
+                break;
+        }
+    }
+
+    if (err == 0) {
+        *ptr = buf;
+    }
+
+    return err;
+}
+
+static int
+getaddrtable(PMIB_IPADDRTABLE *ptr)
+{
+    int err = ERANGE;
+    PMIB_IPADDRTABLE buf = NULL;
+    ULONG bufsz = 0;
+    DWORD ret;
+    size_t i;
+
+    static const size_t max = 2;
+
+    assert(ptr != NULL);
+
+    for (i = 0; err == ERANGE && i < max; i++) {
+        ret = GetIpAddrTable(buf, &bufsz, 0);
+        assert(ret != ERROR_INVALID_PARAMETER &&
+               ret != ERROR_NOT_SUPPORTED);
+        switch (ret) {
+            case ERROR_INSUFFICIENT_BUFFER:
+                err = ERANGE;
+                os_free(buf);
+                if ((buf = (PMIB_IPADDRTABLE)os_malloc(bufsz)) == NULL) {
+                    err = ENOMEM;
+                }
+                break;
+            case ERROR_SUCCESS:
+            default:
+                err = GetLastError();
+                break;
+        }
+    }
+
+    if (err == 0) {
+        *ptr = buf;
+    } else {
+        os_free(buf);
+    }
+
+    return err;
+}
+
+static uint32_t
+getflags(const PIP_ADAPTER_ADDRESSES iface)
+{
+    uint32_t flags = 0;
+
+    if (iface->OperStatus == IfOperStatusUp) {
+        flags |= IFF_UP;
+    }
+    if (iface->IfType == IF_TYPE_SOFTWARE_LOOPBACK) {
+        flags |= IFF_LOOPBACK;
+    }
+    if (!(iface->Flags & IP_ADAPTER_NO_MULTICAST)) {
+        flags |= IFF_MULTICAST;
+    }
+
+    /* FIXME: Shouldn't IFF_LOOPBACK be included here? */
+    switch (iface->IfType) {
+      case IF_TYPE_ETHERNET_CSMACD:
+      case IF_TYPE_IEEE80211:
+      case IF_TYPE_IEEE1394:
+      case IF_TYPE_ISO88025_TOKENRING:
+          flags |= IFF_BROADCAST;
+          break;
+      default:
+          flags |= IFF_POINTTOPOINT;
+          break;
+    }
+
+    return flags;
+}
+
+static int
+copyaddr(
+    os_ifaddrs_t **ifap,
+    const PIP_ADAPTER_ADDRESSES iface,
+    const PMIB_IPADDRTABLE addrtable,
+    const PIP_ADAPTER_UNICAST_ADDRESS addr)
+{
+    int err = 0;
+    int eq = 0;
+    os_ifaddrs_t *ifa;
+    DWORD i;
+    struct sockaddr *sa = (struct sockaddr *)addr->Address.lpSockaddr;
+    size_t size;
+
+    assert(iface != NULL);
+    assert(addrtable != NULL);
+    assert(addr != NULL);
+
+    if ((ifa = os_calloc_s(1, sizeof(*ifa))) == NULL) {
```

Comment:
First and only use of os_calloc in the code base — calloc is standard so I can live with it, but I'm not sure having it in is worth it if it is used so rarely ...

Reply:
It's been available since the initial code submission. I guess nobody found it yet? For now I'd like to leave this in if you don't mind(?)